hydrosdk/deployment_configuration.py

```diff
@@ -0,0 +1,474 @@
+
+    @staticmethod
+    def list(cluster: Cluster):
+        """
+        List all deployment configurations
+
+        :param cluster:
+        :return:
+        """
+        resp = cluster.request("GET", DeploymentConfiguration._BASE_URL)
+        handle_request_error(resp, f"Failed to get a list of Deployment Configurations - {resp.status_code} {resp.text}")
+        return [DeploymentConfiguration.from_camel_case_dict(app_json) for app_json in resp.json()]
+
+    @staticmethod
+    def find(cluster, name):
+        """
+        Search a deployment configuration by name
+
+        :param cluster:
+        :param name:
+        :return:
+        """
+        resp = cluster.request("GET", f"{DeploymentConfiguration._BASE_URL}/{name}")
+        handle_request_error(resp, f"Failed to find Deployment Configuration named {name} {resp.status_code} {resp.text}")
+        return DeploymentConfiguration.from_camel_case_dict(resp.json())
+
+    @staticmethod
+    def delete(cluster, name):
+        """
+        Delete deployment configuration from Hydrosphere cluster
+
+        :param cluster: Active Cluster
+        :param name: Deployment configuration name
+        """
+        resp = cluster.request("DELETE", f"{DeploymentConfiguration._BASE_URL}/{name}")
+        handle_request_error(resp, f"Failed to delete Deployment Configuration named {name}: {resp.status_code} {resp.text}")
+
+
+class DeploymentConfigurationBuilder:
+    """
+    DeploymentConfigBuilder is used to create a new DeploymentConfiguration.
+
+    :Example:
+
+    Create a new deployment configuration.
+    >>> from hydrosdk import Cluster, DeploymentConfigurationBuilder
+    >>> cluster = Cluster('http-cluster-endpoint')
+    >>> config_builder = DeploymentConfigurationBuilder(name="new_config", cluster=cluster)
+    >>> config = config_builder.with_hpa(max_replicas=10, min_replicas=2, target_cpu_utilization_percentage=80). \
+                    with_pod_node_selector({"label1": "key1", "foo": "bar"}). \
+                    with_resource_requirements(limits={"cpu": "4", "memory": "4g"}, \
+                                               requests={"cpu": "2", "memory": "2g"}). \
+                    with_replicas(replica_count=4). \
+                    with_toleration(effect="PreferNoSchedule", key="equalToleration", toleration_seconds=30, operator="Exists").\
+                    build()
+    """
+
+    def __init__(self, name, cluster):
+        """
+        Create new Deployment Config Builder
+
+        :param name: Name of created deployment  configuration
+        :param cluster: Cluster where new deployment configuration should be created
+        """
+        self.name = name
+        self.cluster = cluster
+
+        self.hpa = None
+        self.pod_spec = None
+        self.container_spec = None
+        self.deployment_spec = None
+
+    def with_hpa(self, max_replicas, min_replicas=1, target_cpu_utilization_percentage=None) -> 'DeploymentConfigurationBuilder':
```

Comment:
Aren't with_hpa and with_replicas mutually exclusive? I mean, we should check here, if either of those is already configured.

Reply:
No, it's not. 

AFAIK, k8s scales your deployment based on `replicas` field. HPA doesn't scale your deployment directly, but updates your deployment `replicas` field. Hence, in this situation `replicas` can be seen as initial number of replicas after this definition is applied. 

If `replicas` is omitted, then HPA will scale from 1 replica. If not, HPA will scale from `replicas`.

IMO k8s doesn't throw an exception in this case, so neither should we.

Reply:
👌